misc/releaser.py

```diff
@@ -71,127 +100,215 @@ class ReleaseError(Exception):
             raw = raw[1:]
-        # Git describe (e.g. `-g3b5f5762`) show our position relative to and
-        # existing release, hence this is equivalent to the `+dev` suffix
-        match = re.match(f"^{RELEASE_REGEX}$", raw)
+        if git:
+            # Git describe show our position relative to an existing release.
+            # `git describe` could return for example `1.2.3-15-g3b5f5762`, here:
+            #
+            # - `1.2.3` is the `major.minor.patch`.
+            # - `-15` indicate that we have `15` commit since the tag `1.2.3`.
+            # - `-g3b5f5762` indicate that we are currently on the commit `3b5f5762`.
+            #
+            # A semver compatible version could be `1.2.3-dev.15+git.3b5f5762`
+            #
+            # Convert `1.2.3-N-gSHA` to `1.2.3-dev.N+git.SHA`
+            raw, _ = re.subn(
+                pattern=r"-(\d+)-g([0-9A-Za-z]+)", repl=r"-dev.\1+git.\2", string=raw, count=1
+            )
+        match = RELEASE_REGEX.match(raw)
         if not match:
             raise ValueError(
                 f"Invalid version format {raw!r}, should be `[v]<major>.<minor>.<patch>[-<post>][+dev|-<X>-g<commit>]` (e.g. `v1.0.0`, `1.2.3+dev`, `1.6.7-rc1`)"
             )
 
-        major, minor, patch, pre, dev = match.groups()
-        self.major = int(major)
-        self.minor = int(minor)
-        self.patch = int(patch)
-        self.is_dev = bool(dev)
-        if pre:
-            match = re.match(r"^(a|b|rc)([0-9]+)$", pre)
-            self.pre_type = match.group(1)
-            self.pre_index = int(match.group(2))
-        else:
-            self.pre_type = self.pre_index = None
+        major = int(match.group("major"))
+        minor = int(match.group("minor"))
+        patch = int(match.group("patch"))
+        prerelease = match.group("pre")
+        dev = int(match.group("dev")) if match.group("dev") is not None else None
+        local = match.group("local")
 
-    def evolve(self, **kwargs):
+        return Version(major, minor, patch, prerelease=prerelease, dev=dev, local=local)
+
+    def evolve(self, **kwargs: Any) -> Version:
         new = copy(self)
         new.__dict__.update(kwargs)
         return new
 
-    def __str__(self):
-        base = f"v{self.major}.{self.minor}.{self.patch}"
-        if self.is_preversion:
-            base += f"-{self.pre_type}{self.pre_index}"
-        if self.is_dev:
-            base += "+dev"
+    def with_v_prefix(self) -> str:
+        return "v" + str(self)
+
+    def __repr__(self) -> str:
+        return f"Version(major={self.major}, minor={self.minor}, patch={self.patch}, prerelease={self.prerelease}, dev={self.dev}, local={self.local})"
+
+    def __str__(self) -> str:
+        base = f"{self.major}.{self.minor}.{self.patch}"
+        if self.pre is not None:
+            type, index = self.pre
+            base += f"-{type}{index}"
+        if self.dev is not None:
+            base += "-dev" + f".{self.dev}" if self.dev else ""
```

Comment:
```suggestion
            base += f"-dev.{self.dev}" if self.dev else ""
```

Reply:
I've partially applied the suggestion:

The `dev` part may contain a number after it but that value isn't required